wal/wal.go

```diff
@@ -834,3 +834,17 @@ func (r *segmentBufReader) Read(b []byte) (n int, err error) {
 	r.buf.Reset(r.segs[r.cur])
 	return n, nil
 }
+
+// Computing size of the WAL.
+// We do this by counting the number of segments currently used by the WAL,
+// and multiplying it by the size of a segment.
+func (w *WAL) Size() (size int64, err error) {
+	first, last, err := w.Segments()
+	if err != nil {
+		return 0, err
+	}
+	if first == -1 || last == -1 {
+		return 0, fmt.Errorf("no segments found in WAL")
+	}
+	return int64((last - first + 1) * w.segmentSize), nil
```

Comment:
what about if the last segment is not complete?

Reply:
As far as I understand, the `WAL.Segments()` call returns the indices of the first and last segments on disk. Considering the fact that segment files are created on disk when they're initialised, means that the function will return the index of the current, incomplete segment as `last`.

Reply:
that is right and the last segment will not be the full  w.segmentSize

Reply:
So I do not need to modify my WAL size calculations right?

Reply:
No I mean here you assume that all segments are of size `w.segmentSize`, but in reality the last segment is not.

Reply:
Okay, so a more accurate calculation would be something like `((last - first) * w.segmentSize) + (w.donePages * pageSize)`?

Reply:
Segments don't have to be the full size.

Reply:
Yeah maybe something like that, I am not sure, need to look at the code.

Double check if this might cause a race condition somewhere.

Reply:
If segments are not necessarily completely filled in before being switched out, at that point would it not be easier to read the segment files directly from the `data/wal/` directory, and adding up the byte totals? Something like:

```
	files, err := ioutil.ReadDir(w.Dir())
	if err != nil {
		return 0, err
	}
	for _, fi := range files {
		size += fi.Size()
	}
```
This won't take into account the current page that has not been flushed to disk, though.

Reply:
You'd need that anyway for checkpoints.

Reply:
@brian-brazil Are you referring to WAL checkpoints?

Reply:
Yes

Reply:
Okay, I've modified the code to now read the contents of the `data/wal/` directory and add the sizes of all the files, including the checkpoints.

Reply:
I preffered the other way , just need to figure out how to calculate it for the active segment.

Reply:
When calculating the WAL size using the segments, it too scans the WAL dir and returns the names of the segment files, the size of which we assume to be `w.segmentSize`, but as @brian-brazil said, its not necessary that each of the segments is fully occupied. This way we get the true occupied size on disk, which would be more accurate, and we wouldn't have to account for the active segments' pages sizes either.

Reply:
@krasi-georgiev Were you able to take a look at an alternative approach?

Reply:
not yet, probably next week. I haven't forgotten , just need to finish some other things I have started.